xwayland/xwm.c

```diff
@@ -492,84 +614,282 @@ static void handle_configure_request(struct wlr_xwm *xwm,
+	struct wlr_xwayland_surface *xsurface = lookup_surface(xwm, client_message->window);
+	if (!xsurface) {
+		return;
+	}
+
+	int maximized = xsurface_is_maximized(xsurface);
+
+	uint32_t action = client_message->data.data32[0];
+	uint32_t property = client_message->data.data32[1];
+
+	if (property == xwm->atoms[_NET_WM_STATE_FULLSCREEN] &&
+			update_state(action, &xsurface->fullscreen)) {
+		xsurface_set_net_wm_state(xsurface);
+
+		if (xsurface->fullscreen) {
+			xsurface->saved_width = xsurface->width;
+			xsurface->saved_height = xsurface->height;
+		}
+
+		wl_signal_emit(&xsurface->events.request_fullscreen, xsurface);
+	} else {
+		if (property == xwm->atoms[_NET_WM_STATE_MAXIMIZED_VERT] &&
+				update_state(action, &xsurface->maximized_vert)) {
+			xsurface_set_net_wm_state(xsurface);
+		}
+
+		if (property == xwm->atoms[_NET_WM_STATE_MAXIMIZED_HORZ] &&
+				update_state(action, &xsurface->maximized_horz)) {
+			xsurface_set_net_wm_state(xsurface);
+		}
+
+		if (maximized != xsurface_is_maximized(xsurface)) {
+			if (xsurface_is_maximized(xsurface)) {
+				xsurface->saved_width = xsurface->width;
+				xsurface->saved_height = xsurface->height;
+			}
+
+			wl_signal_emit(&xsurface->events.request_maximize, xsurface);
+		}
+	}
+}
+
 static void handle_client_message(struct wlr_xwm *xwm,
 		xcb_client_message_event_t *ev) {
 	wlr_log(L_DEBUG, "XCB_CLIENT_MESSAGE (%u)", ev->window);
 
 	if (ev->type == xwm->atoms[WL_SURFACE_ID]) {
-		struct wlr_xwayland_surface *surface = lookup_surface(
-			&xwm->new_surfaces, ev->window);
-		if (surface == NULL) {
-			wlr_log(L_DEBUG, "client message WL_SURFACE_ID but no new window %u ?",
-				ev->window);
-			return;
-		}
-		surface->surface_id = ev->data.data32[0];
-		/* Check if we got notified after wayland surface create event */
-		struct wl_resource *resource = wl_client_get_object(
-			xwm->xwayland->client, surface->surface_id);
-		if (resource) {
-			map_shell_surface(xwm, surface, wl_resource_get_user_data(resource));
-		} else {
-			wl_list_remove(&surface->link);
-			wl_list_insert(&xwm->unpaired_surfaces, &surface->link);
-		}
+		handle_surface_id_message(xwm, ev);
 	} else if (ev->type == xwm->atoms[NET_WM_STATE]) {
-		struct wlr_xwayland_surface *surface = lookup_surface_any(xwm,
-			ev->window);
-		if (surface == NULL) {
-			return;
-		}
-		handle_surface_state(xwm, surface, &ev->data.data32[1], 2,
-			ev->data.data32[0]);
+		handle_net_wm_state_message(xwm, ev);
+	} else if (ev->type == xwm->atoms[_NET_WM_MOVERESIZE]) {
+		handle_net_wm_moveresize_message(xwm, ev);
 	} else {
 		wlr_log(L_DEBUG, "unhandled x11 client message %u", ev->type);
 	}
 }
 
+static void handle_focus_in(struct wlr_xwm *xwm,
+		xcb_focus_in_event_t *ev) {
+	// Do not interfere with grabs
+	if (ev->mode == XCB_NOTIFY_MODE_GRAB ||
+			ev->mode == XCB_NOTIFY_MODE_UNGRAB) {
+		return;
+	}
+
+	// Do not let X clients change the focus behind the compositor's
+	// back. Reset the focus to the old one if it changed.
```

Comment:
FWIW this may not be desirable. Things like dmenu don't work well with this logic. It's a tricky question.

Reply:
Do you have a specific issue?

Reply:
No, just something to think about.